File: rain_framework/src/executioner.rs
use godot::prelude::*;
use godot::classes::{INode, Node};
use godot::classes::Engine;

use crate::entity::Entity;
use crate::hit_event::*;

#[derive(GodotClass)]
#[class(tool, base=Node)]
pub struct Executioner
{
    #[var]
    verbose: bool,
    events: Array<Option<Gd<HitEvent>>>,
    base: Base<Node>
}

#[godot_api]
impl INode for Executioner
{
    fn init(base:Base<Node>) -> Self
    {
        Self {
            verbose: false,
            events: Array::new(),
            base
        }
    }

    fn enter_tree(&mut self)
    {
        
    }

    fn physics_process(&mut self, _delta: f64)
    {
        if Engine::singleton().is_editor_hint() { return }
        
        if !self.events.is_empty()
        {
            self.ev_handler();
        }
    }
}

#[godot_api]
impl Executioner
{
    fn ev_handler(&mut self)
    {
        for event in self.events.iter_shared()
        {
            if let mut ev = event.unwrap()
            {
                let ev_type = ev.bind().event_type;
                let force = ev.bind().get_force();
                let mut from = ev.bind_mut().get_from().unwrap();
                let mut to = ev.bind_mut().get_to().unwrap();

                match ev_type
                {
                    // Hit
                    EventType::HIT => {
                        let velocity = to.get_velocity();
                        to.set_velocity(velocity + force);
                    },
                    // Parry
                    EventType::PARRY => {
                        
                    },
                    // Shield
                    EventType::SHIELD => {

                    },
                    // Evade
                    EventType::EVADE => {

                    },
                    // Bump
                    EventType::BUMP => {

                    },
                    // Grabbed
                    EventType::GRABBED => {

                    }
                }
            }
        }

    }
}

File: detonator_2/cs_classes/global/Executioner.cs
using Godot;
using Godot.Collections;
using System;

public partial class Executioner : Node
{
    private long index = -1;
    [Export] private Dictionary<long, HitEvent> events = new Dictionary<long, HitEvent>();

    public override void _PhysicsProcess(double _delta)
    {
        foreach (HitEvent ev in events.Values)
            hit_event_handler(ev);

        index = -1;
    }

    private HitEvent create_event(HitEvent.EventType type, Node from, Node to, Dictionary values)
    {
        return new HitEvent(type, from, to, values);
    }

    public void add_event(HitEvent.EventType type, Node from, Node to, Dictionary values)
    {
        index += 1;
        var ev = create_event(type, from, to, values);
        events.Add(index, ev);
    }

    public void hit_event_handler(HitEvent ev)
    {
        switch (ev.event_type)
        {
            case HitEvent.EventType.HIT:
                break;
            case HitEvent.EventType.SHIELD:
                break;
            case HitEvent.EventType.EVADE:
                break;
            case HitEvent.EventType.GRABBED:
                break;
            case HitEvent.EventType.PARRY:
                break;
        }
    }

    private void hit(Node from, Node to, float force, Vector2 dir)
    {
        
    }

}
